### PR TITLE
JaegerExporter: binary thrift over http

### PR DESCRIPTION
**Pull request by CodeBlanch**
## Changes

* Adding support for the binary thrift protocol over http to jaeger collector.
* Cleaned up and simplified things. Now there is only 1) a protocol (Compact or Binary) that writes to a buffer and 2) a client (UDP or HTTP) which sends the buffers. Previously we had protocols, transports, & 2 types of clients (thrift client & jaeger client).

## Public API Additions

```csharp
    public enum JaegerExportProtocol : byte
    {
        UdpCompactThrift = 0,

        HttpBinaryThrift = 1,
    }

    public JaegerExporterOptions()
    {
	public JaegerExportProtocol Protocol { get; set; } = JaegerExportProtocol.UdpCompactThrift;

	public Uri Endpoint { get; set; } = new Uri("http://localhost:14268");

	public Func<HttpClient> HttpClientFactory { get; set; }
    }
```

## Benchmarks

Code is a little smarter now about how it uses buffers. We pre-calculate the batch header & footer. The header just lives in the buffer forever. The footer is copied in a block at the end after the spans. The bytes for seqId + span count are updated in the header before transmission since they aren't known initially. Spans are first written to a temp buffer and then written into the batch buffer if there is room. The result is there is less code to execute, less copying of data, and the result is: Faster processing!

### Before

|                  Method | NumberOfBatches | NumberOfSpans |        Mean |     Error |    StdDev | Allocated |
|------------------------ |---------------- |-------------- |------------:|----------:|----------:|----------:|
| JaegerExporter_Batching |               1 |         10000 |    71.03 ms |  0.696 ms |  0.651 ms |     21 KB |
| JaegerExporter_Batching |              10 |         10000 |   722.48 ms |  4.290 ms |  3.803 ms |     23 KB |
| JaegerExporter_Batching |             100 |         10000 | 7,531.84 ms | 58.356 ms | 54.586 ms |     22 KB |

### After

|                  Method | NumberOfBatches | NumberOfSpans |        Mean |     Error |    StdDev | Allocated |
|------------------------ |---------------- |-------------- |------------:|----------:|----------:|----------:|
| JaegerExporter_Batching |               1 |         10000 |    **55.99 ms** |  0.365 ms |  0.342 ms |     18 KB |
| JaegerExporter_Batching |              10 |         10000 |   **541.78 ms** |  5.070 ms |  4.234 ms |     19 KB |
| JaegerExporter_Batching |             100 |         10000 | **5,222.58 ms** | 46.481 ms | 41.204 ms |     19 KB |

Note: Memory is only lower because I used `this.maxPayloadSizeInBytes` to calculate the initial buffer size. The main thing is that it remains flat.

## TODOs

* [X] `CHANGELOG.md` updated for non-trivial changes
* [X] Unit tests
* [X] Benchmarks
* [X] Documentation
* [ ] Changes in public API reviewed

### CODE REVIEW

**Comment by ElectricVampire** [comment]
@CodeBlanch Thanks for all this work, I was also doing the similar thing (took inspiration from Java implementation of this library) and noticed you already have a PR for it. This is must for me as currently I am using Legacy Zipkin port in Jaeger for Http Post and Jaeger marked it as deprecated. Do you have any estimate for this that when you will be able to push it.
Let me know if there is any thing where I could help or assist.
Again thank you very much for this work.

**Comment by CodeBlanch** [comment]
@ElectricVampire No problem! It is a big one so I'm chipping away at it a little bit each day. I have a TODO list in the description if you want to follow along. Just need to do some documentation now. I expect in a day or two to wrap that up and convert it into a real PR. Then we need to get some review. A lot of the team is busy trying to finish metrics right now so I'm not sure how long the review will take. Probably not that long, I think a week is a conservative guess. But then the hard part, we need to get it released. That will probably take the most time because of the metrics work. Maybe late November? Check out: https://github.com/open-telemetry/opentelemetry-dotnet/milestones

**Comment by ElectricVampire** [comment]
> @ElectricVampire No problem! It is a big one so I'm chipping away at it a little bit each day. I have a TODO list in the description if you want to follow along. Just need to do some documentation now. I expect in a day or two to wrap that up and convert it into a real PR. Then we need to get some review. A lot of the team is busy trying to finish metrics right now so I'm not sure how long the review will take. Probably not that long, I think a week is a conservative guess. But then the hard part, we need to get it released. That will probably take the most time because of the metrics work. Maybe late November? Check out: https://github.com/open-telemetry/opentelemetry-dotnet/milestones

Great, I'll monitor this PR and To-Do list. Most of the work is already done and seems better than my current solution which mimics java library. If something major comes after the review where I can contribute, I'll try to do it. 

This will be great addition as this will allow many software solutions to do seem less transformation from OpenTracing to OpenTelemetry.

File: src/OpenTelemetry.Exporter.Jaeger/Implementation/JaegerHttpClient.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    {
+        private static readonly MediaTypeHeaderValue ContentTypeHeader = new MediaTypeHeaderValue("application/vnd.apache.thrift.binary");
+
+        private readonly Uri endpoint;
+        private readonly HttpClient httpClient;
+        private bool disposed;
+
+        public JaegerHttpClient(Uri endpoint, HttpClient httpClient)
+        {
+            this.endpoint = endpoint ?? throw new ArgumentNullException(nameof(endpoint));
```
Comment (reyang) [inline src/OpenTelemetry.Exporter.Jaeger/Implementation/JaegerHttpClient.cs:33]: Consider `Debug.Assert` given the class is internal.

File: src/OpenTelemetry.Exporter.Jaeger/Implementation/JaegerUdpClient.cs [hunk 2]
```diff
@@ -50,28 +45,18 @@ public int Send(byte[] buffer, int offset, int count)
...
-            {
-                this.client.Dispose();
-            }
+            this.client.Dispose();
 
             this.disposed = true;
         }
+
+        public override string ToString()
+            => $"{this.host}:{this.port}";
```
Comment (reyang) [inline src/OpenTelemetry.Exporter.Jaeger/Implementation/JaegerUdpClient.cs:60]: Is this for debugging purpose? (it's a bit weird, maybe consider `JaegerUdpClient("udp://host:port")`?)

File: src/OpenTelemetry.Exporter.Jaeger/JaegerExporterHelperExtensions.cs [hunk 3]
```diff
@@ -40,17 +41,55 @@ public static TracerProviderBuilder AddJaegerExporter(this TracerProviderBuilder
...
         {
             configure?.Invoke(options);
 
+            if (options.Protocol == JaegerExportProtocol.HttpBinaryThrift && options.HttpClientFactory == null)
+            {
+                if (serviceProvider != null)
+                {
+                    options.HttpClientFactory = () =>
+                    {
+                        Type httpClientFactoryType = Type.GetType("System.Net.Http.IHttpClientFactory, Microsoft.Extensions.Http", throwOnError: false);
```
Comment (reyang) [inline src/OpenTelemetry.Exporter.Jaeger/JaegerExporterHelperExtensions.cs:65]: What's the typical number of invocations expected for this factory? If more than once, perhaps extract the reflection logic out.

File: src/OpenTelemetry.Exporter.Jaeger/README.md [hunk 4]
```diff
@@ -37,16 +37,28 @@ take precedence over the environment variables.
...
+  | Protocol       | Description                   |
+  |----------------|-------------------------------|
+  |UdpCompactThrift| Apache Thrift compact over UDP to a Jaeger Agent.     |
+  |HttpBinaryThrift| Apache Thrift binary over HTTP to a Jaeger Collector. |
+
 * `AgentHost`: The Jaeger Agent host (default `localhost`).
-* `AgentPort`: The compact thrift protocol UDP port of the Jaeger Agent
-  (default `6831`).
+* `AgentPort`: The Jaeger Agent port (default `6831`).
+* `Endpoint`: The Jaeger Collector HTTP endpoint (default
```
Comment (reyang) [inline src/OpenTelemetry.Exporter.Jaeger/README.md:49]: nit: consider sort alphabetically

File: src/OpenTelemetry.Exporter.Jaeger/README.md [hunk 5]
```diff
@@ -59,12 +71,46 @@ values of the `JaegerExporterOptions`.
 
 | Environment variable              | `JaegerExporterOptions` property |
 | --------------------------------- | -------------------------------- |
+| `OTEL_EXPORTER_JAEGER_PROTOCOL`   | `Protocol`                      |
 | `OTEL_EXPORTER_JAEGER_AGENT_HOST` | `AgentHost`                      |
 | `OTEL_EXPORTER_JAEGER_AGENT_PORT` | `AgentPort`                      |
+| `OTEL_EXPORTER_JAEGER_ENDPOINT` | `Endpoint`                      |
```
Comment (reyang) [inline src/OpenTelemetry.Exporter.Jaeger/README.md:77]:
```suggestion
| `OTEL_EXPORTER_JAEGER_ENDPOINT`   | `Endpoint`                       |
```

**Review by reyang — APPROVED** [review]

File: src/OpenTelemetry.Exporter.Jaeger/JaegerExporterHelperExtensions.cs (hunk 3 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Jaeger/JaegerExporterHelperExtensions.cs:65]: Just once during construction of the exporter instance: https://github.com/open-telemetry/opentelemetry-dotnet/blob/5fd75ba23eb43049dae0d31dd109088fbef7e18f/src/OpenTelemetry.Exporter.Jaeger/JaegerExporter.cs#L70

File: src/OpenTelemetry.Exporter.Jaeger/Implementation/JaegerUdpClient.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Jaeger/Implementation/JaegerUdpClient.cs:60]: I think it is just there for debugging. I'll remove it.

File: src/OpenTelemetry.Exporter.Jaeger/README.md (hunk 4 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Jaeger/README.md:49]: Updated

File: src/OpenTelemetry.Exporter.Jaeger/README.md (hunk 5 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.Jaeger/README.md]: Updated

**Pull request merged**